lib/Sisimai/Rhost/ExchangeOnline.pm

```diff
@@ -4,168 +4,723 @@ use strict;
-        ],
-        qr/\A4[.]7[.][568]\d\d\z/ => [
-            { 'reason' => 'securityerror', 'string' => ['Access denied, please try again later'] }
-        ],
-        qr/\A5[.]1[.][07]\z/ => [
-            { 'reason' => 'rejected', 'string' => ['Sender denied', 'Invalid address'] }
-        ],
-        qr/\A5[.]1[.][123]\z/ => [{
-            'reason' => 'userunknown',
-            'string' => [
-                'Bad destination mailbox address',
-                'Invalid X.400 address',
-                'Invalid recipient address',
-            ]
-        }],
-        qr/\A5[.]4[.][46]\z/ => [{
-            'reason' => 'networkerror',
-            'string' => ['Invalid arguments', 'Routing loop detected'],
-        }],
-        qr/\A5[.]7[.][13]\z/ => [{
-            'reason' => 'securityerror',
-            'string' => ['Delivery not authorized', 'Not Authorized'],
-        }],
-        qr/\A5[.]7[.]50[1-3]\z/ => [{
-            'reason' => 'spamdetected',
-            'string' => [
-                'Access denied, spam abuse detected',
-                'Access denied, banned sender'
-            ],
-        }],
-        qr/\A5[.]7[.]50[457]\z/ => [{
-            'reason' => 'filtered',
-            'string' => [
-                'Recipient address rejected: Access denied',
-                'Access denied, banned recipient',
-                'Access denied, rejected by recipient'
-            ]
-        }],
-        qr/\A5[.]7[.]6\d\d\z/ => [
-            { 'reason' => 'blocked', 'string' => ['Access denied, banned sending IP '] }
-        ],
-        qr/\A5[.]7[.]7\d\d\z/ => [
-            { 'reason' => 'toomanyconn', 'string' => ['Access denied, tenant has exceeded threshold'] }
-        ],
-    };
     state $messagesof = {
-        # Copied and converted from Sisimai::Lhost::Exchange2007
-        'expired'       => ['QUEUE.Expired'],
-        'hostunknown'   => ['SMTPSEND.DNS.NonExistentDomain'],
-        'mesgtoobig'    => ['RESOLVER.RST.RecipSizeLimit', 'RESOLVER.RST.RecipientSizeLimit'],
-        'networkerror'  => ['SMTPSEND.DNS.MxLoopback'],
-        'rejected'      => ['RESOLVER.RST.NotAuthorized'],
-        'securityerror' => ['RESOLVER.RST.AuthRequired'],
-        'systemerror'   => [
-            'RESOLVER.ADR.Ambiguous',
-            'RESOLVER.ADR.BadPrimary',
-            'RESOLVER.ADR.InvalidInSmtp',
-            'RESOLVER.FWD.NotFound',
-        ],
-        'toomanyconn'   => ['RESOLVER.ADR.RecipLimit', 'RESOLVER.ADR.RecipientLimit'],
-        'userunknown'   => [
-            'RESOLVER.ADR.RecipNotFound',
-            'RESOLVER.ADR.RecipientNotFound',
-            'RESOLVER.ADR.ExRecipNotFound',
-            'RESOLVER.ADR.ExRecipientNotFound',
+        'authfailure' => [
+            # - Access denied, a message sent over IPv6 [2a01:111:f200:2004::240] must pass either
+            #   SPF or DKIM validation, this message is not signed
+            # - The sending message sent over IPv6 must pass either SPF or DKIM.
```

Comment:
# [ "Error Code", minimum of the 3rd value, maximum of the 3rd value, "Error Message"]